aws_config_test.go

```diff
@@ -1185,6 +1185,170 @@ func fullValueTypeName(v reflect.Value) string {
 	return fmt.Sprintf("%s.%s", requestType.PkgPath(), requestType.Name())
 }
 
+func TestRegion(t *testing.T) {
+	testCases := map[string]struct {
+		Config                  *Config
+		EnvironmentVariables    map[string]string
+		SharedConfigurationFile string
+		ExpectedRegion          string
+	}{
+		"no configuration": {
+			Config: &Config{
+				AccessKey: servicemocks.MockStaticAccessKey,
+				SecretKey: servicemocks.MockStaticSecretKey,
+			},
+			ExpectedRegion: "",
+		},
+
+		"config": {
+			Config: &Config{
+				AccessKey: servicemocks.MockStaticAccessKey,
+				Region:    "us-east-1",
+				SecretKey: servicemocks.MockStaticSecretKey,
+			},
+			ExpectedRegion: "us-east-1",
+		},
+
+		"AWS_REGION": {
+			Config: &Config{
+				AccessKey: servicemocks.MockStaticAccessKey,
+				SecretKey: servicemocks.MockStaticSecretKey,
+			},
+			EnvironmentVariables: map[string]string{
+				"AWS_REGION": "us-east-1",
+			},
+			ExpectedRegion: "us-east-1",
+		},
+		"AWS_DEFAULT_REGION": {
+			Config: &Config{
+				AccessKey: servicemocks.MockStaticAccessKey,
+				SecretKey: servicemocks.MockStaticSecretKey,
+			},
+			EnvironmentVariables: map[string]string{
+				"AWS_DEFAULT_REGION": "us-east-1",
+			},
+			ExpectedRegion: "us-east-1",
+		},
+		"AWS_REGION overrides AWS_DEFAULT_REGION": {
+			Config: &Config{
+				AccessKey: servicemocks.MockStaticAccessKey,
+				SecretKey: servicemocks.MockStaticSecretKey,
+			},
+			EnvironmentVariables: map[string]string{
+				"AWS_REGION":         "us-east-1",
+				"AWS_DEFAULT_REGION": "us-west-2",
+			},
+			ExpectedRegion: "us-east-1",
+		},
+
+		"shared configuration file": {
+			Config: &Config{
+				AccessKey: servicemocks.MockStaticAccessKey,
+				SecretKey: servicemocks.MockStaticSecretKey,
+			},
+			SharedConfigurationFile: `
+[default]
+region = us-east-1
+`,
+			ExpectedRegion: "us-east-1",
+		},
+
+		"config overrides AWS_REGION": {
```

Comment:
so by transitive property (🤣 ) then config overrides shared credentials too?

Reply:
Exactly. The exercise was left to the reader… 😉